implementations/rust/ockam/ockam_core/src/bare.rs

```diff
@@ -0,0 +1,226 @@
+//! Primitives to encode and decode Binary Application Record Encoding (BARE).
+//!
+//! These primitives are used to encode and decode most performance-sensitive messages
+//! in Ockam without using `std`, and not are not meant to support all possible use cases.
+
+/// Defines a set of functions to encode and decode bare encoding
+///
+/// This module is not dependent on std or any other crate
+use crate::compat::vec::Vec;
+
+/// Read a dynamically sized slize from the given cursor
+pub fn read_slice<'de>(slice: &'de [u8], index: &mut usize) -> Option<&'de [u8]> {
+    let length: usize = read_variable_length_integer(slice, index)?
+        .try_into()
+        .ok()?;
+    if slice.len() - *index >= length {
+        let result = &slice[*index..(*index + length)];
+        *index += length;
+        Some(result)
+    } else {
+        None
+    }
+}
+
+/// Returns the size of the encoded slice in bytes
+pub fn size_of_slice(slice: &[u8]) -> usize {
+    size_of_variable_length(slice.len() as u64) + slice.len()
+}
+
+/// Write a dynamically sized slice to the given buffer
+pub fn write_slice(destination: &mut Vec<u8>, buffer: &[u8]) {
+    write_variable_length_integer(destination, buffer.len() as u64);
+    destination.extend_from_slice(buffer);
+}
+
+/// Reads a string from the given cursor
+pub fn read_str<'de>(slice: &'de [u8], index: &mut usize) -> Option<&'de str> {
+    let buffer = read_slice(slice, index)?;
+    core::str::from_utf8(buffer).ok()
+}
+
+/// Writes a string to the given buffer
+pub fn write_str(destination: &mut Vec<u8>, string: &str) {
+    write_slice(destination, string.as_bytes());
+}
+
+/// Returns the size in bytes of the given variable length integer
+pub fn size_of_variable_length(value: u64) -> usize {
+    let mut result = 0;
+    let mut value = value;
+    loop {
+        value >>= 7;
+        result += 1;
+        if value == 0 {
+            break;
+        }
+    }
+    result
+}
+
+/// Read a variable length integer from the given cursor (ULEB128)
+/// returns None if the buffer is too short
+pub fn read_variable_length_integer(slice: &[u8], index: &mut usize) -> Option<u64> {
+    let mut result = 0;
+    let mut shift = 0;
+    loop {
+        let byte = slice.get(*index)?;
+        *index += 1;
+
+        result |= ((byte & 0b0111_1111) as u64) << shift;
+        if byte & 0b1000_0000 == 0 {
+            break;
+        }
+        shift += 7;
+
+        if shift >= 64 {
+            return None;
+        }
+    }
+
+    Some(result)
+}
+
+/// Write a variable length integer to the given buffer (ULEB128)
+pub fn write_variable_length_integer(destination: &mut Vec<u8>, mut value: u64) {
```

Comment:
Just in case you like, you can do this way:
```rs
pub fn write_variable_length_integer(destination: &mut Vec<u8>, mut value: u64) {
    loop {
        let mut byte = (value & 0b0111_1111) as u8;
        value >>= 7;
        if value != 0 {
            destination.push(byte | 0b1000_0000);
        } else {
            destination.push(byte);
            break;
        }
    }
}
```